### ISSUE TITLE
Anagram: Test simplification and/or removal

### ISSUE THREAD

**Issue by Zhiyuan-Amos** [exercism/problem-specifications#1067]
I think that the suite of tests in `Anagram` can be improved.

1. Some of the tests are already covered by other tests and can be removed.
2. Some tests are testing for too many things and can be simplified.

I've opened a PR in [exercism/java](https://github.com/exercism/java/pull/1164), please do take a look and see if what I'm doing makes sense :)

**Comment by rpottsoh** [comment]
test data: https://github.com/exercism/problem-specifications/blob/master/exercises/anagram/canonical-data.json

test description: https://github.com/exercism/problem-specifications/blob/master/exercises/anagram/description.md

@Zhiyuan-Amos it is nice to include links to the relevant files that define the exercise to make it easier for everyone to quickly look up the relevant files that define the exercise.  There are many exercises here and I don't know that everyone, *or anyone*, is intimately familiar with all of them.  :smile:

**Comment by rpottsoh** [comment]
1. It might make things easier here if you would call out which tests you think should be removed and why.  This would be easier than trying to jump over to the PR in the Java track.
1. Which tests so you think are testing too many things and should be simplified.

In otherwords get everything out in the open here and likely a discussion will ensue.  As you were forewarned in exercism/java#1164 the discussion can sometimes take awhile as everyone here is a volunteer and a lot will also depend on how controversial people find the issue you have raised.

**Comment by Zhiyuan-Amos** [comment]
```json
{
  "description": "does not detect identical words",
  "property": "anagrams",
  "input": {
    "subject": "corn",
    "candidates": ["corn", "dark", "Corn", "rank", "CORN", "cron", "park"]
  },
  "expected": ["cron"]
}
```

This test case is too complex. The candidates should only contain "corn", "Corn", "CORN", since it is simply testing that identical words of different casings are not being considered as anagrams. Also, we can increase the complexity of this test case by updating the subject to be mixed-case, such as "cOrN".

```json
{
  "description": "does not detect a anagram if the original word is repeated",
  "property": "anagrams",
  "input": {
    "subject": "go",
    "candidates": ["go Go GO"]
  },
  "expected": []
}
```

If we are updating the previous test case mentioned above, then this test case is actually the same as the previous and can be removed. Even if we aren't updating the previous test case, both test cases are actually very similar, and we can consider removing one of them.

```json
{
  "description": "anagrams must use all letters exactly once",
  "property": "anagrams",
  "input": {
    "subject": "tapper",
    "candidates": ["patter"]
  },
  "expected": []
}

{
  "description": "does not detect false positives",
  "property": "anagrams",
  "input": {
    "subject": "galea",
    "candidates": ["eagle"]
  },
  "expected": []
}
```

Both tests seem to be both testing for "anagrams must use all letters exactly once".

```json
{
  "description": "detects anagrams case-insensitively",
  "property": "anagrams",
  "input": {
    "subject": "Orchestra",
    "candidates": ["cashregister", "Carthorse", "radishes"]
  },
  "expected": ["Carthorse"]
},
{
  "description": "detects anagrams using case-insensitive subject",
  "property": "anagrams",
  "input": {
    "subject": "Orchestra",
    "candidates": ["cashregister", "carthorse", "radishes"]
  },
  "expected": ["carthorse"]
},
{
  "description": "detects anagrams using case-insensitive possible matches",
  "property": "anagrams",
  "input": {
    "subject": "orchestra",
    "candidates": ["cashregister", "Carthorse", "radishes"]
  },
  "expected": ["Carthorse"]
}
```

These 3 tests are testing that anagrams are detected when:
1. Only the subject begins with uppercase letter
2. Only the candidate begins with uppercase letter
3. Both the subject & candidate begins with uppercase letter

These 3 tests can be combined into a higher-level difficulty test case, such that we can have both subject & candidate with mixed-cased letters. E.g. subject: "OrChEsTra", candidates: "carthorse", "CarThorsE", "CARTHORSE". Note that this test case is actually very similar to the test case "does not detect identical words" (right all the way on top).

**Comment by Zhiyuan-Amos** [comment]
@rpottsoh seeking your assistance to take a look at this issue whenever you are free. tqvm! :)

**Comment by rpottsoh** [comment]
@Zhiyuan-Amos thanks for the bump :+1:   I will try to review what you have posted over the coming days.  In the mean time hopefully one of the other maintainers that frequent these issues will have a chance to weigh in on what you have brought up.

**Comment by coriolinus** [comment]
@Zhiyuan-Amos Thanks for writing up the detailed descriptions of your improvements. 

1. `subject: corn`: I agree with you that this test considers more than one property.
2. `subject: go`: If we remove one of the two test cases, I'd remove the `corn` case and keep the `go` case; that 
3. `subject: tapper` and `subject: galea`: Agree that these two test fundamentally the same property, and one might be removed without appreciably changing test coverage.
4. `subject: orchestra`: disagree that merging these tests would be beneficial. While you're correct in saying that we wouldn't likely lose anything by replacing these three with a single case with mixed capitalization, the tests as we have them now test each proposition independently. I believe this to be better test design.

Despite agreeing that your proposed changes do what you suggest, I'd recommend against reducing the number of tests as you propose. In the general case, and definitely in the specific cases considered here, unit tests have essentially zero cost to run. As such, there's no penalty associated with having a large number of them. 

Conversely, there _is_ a penalty associated with writing a smaller number of more complicated tests: they're harder for students to understand. Students progress through exercism's exercises by reading the tests, inferring the design implied, and then writing the necessary code to implement that design. Keeping things simple helps them figure out the intended featureset with a minimum of friction.

Additionally, there's always the risk of accidentally dropping an intended feature when combining a number of tests into a single more complicated test. I don't believe that we'd run into that issue here, but it's something which needs to be watched out for. It's much simpler and less risky to simply never remove a test.

In short: while I'm always willing to consider adding or reordering some tests to clarify an exercise for the student, I have some resistance to merging or deleting tests, even when they are in fact redundant.

**Comment by Zhiyuan-Amos** [comment]
@coriolinus Ah yes, I agree with you that writing a smaller number of more complicated tests are harder for students to understand. As such, it seems that merging smaller tests into a single complicated test isn't the way to go.

However, I think that there's actually a cost of having many tests where some of them are redundant. 
1. It makes it harder for maintainers to understand the existing test class: whether all possible cases have been covered and what each test case is specifically testing.
1. Keeping the tests concise for the students result in them having to read lesser test data. 

> subject: orchestra: disagree that merging these tests would be beneficial. While you're correct in saying that we wouldn't likely lose anything by replacing these three with a single case with mixed capitalization, the tests as we have them now test each proposition independently. I believe this to be better test design.

Ah, yes I agree with you. I'm thinking of the value of these 3 test cases though because only the first letter is being tested. Perhaps we should add an additional test case with mixed capitalization?

Thank you for giving such a thorough comment :)

**Comment by rpottsoh** [comment]
I don't really have anything to add.  @Zhiyuan-Amos and @coriolinus have provided detailed insights.  I am not sure what the next step should be.  @Zhiyuan-Amos you might consider opening a PR that contains the new test case you think might be a good addition and reference this issue.
